wit/request-reply.wit

```diff
@@ -0,0 +1,27 @@
+/// The request-reply interface allows a guest to send a message and await a response. This
+/// interface is considered optional as not all message services support the concept of
+/// request/reply. However, request/reply is a very common pattern in messaging and as such, we have
+/// included it as a core interface.
+interface request-reply {
+    use types.{client, message, error};
+
+    /// Options for a request/reply operation
+    record request-options {
+        /// The maximum amount of time to wait for a response. If the timeout value is not set, then
+        /// the request/reply operation will block until a message is received in response.
+        timeout-ms: option<u32>,
+        /// The number of expected replies. If the number of replies is not set, then the
```

Comment:
@danbugs @devigned As I was adding the new request options, I thought about how we want to handle this. Currently, `request` returns a list of messages. If we are returning a list of messages, then it makes sense we might want to specify (optionally) the number of replies we expect. I am not totally sold on the idea, but put it in here so it was easy to grok. I think there are some options here:

- Keep support for receiving multiple responses and have this optional `expected-replies` field (more complex, but more flexible in what you can use it for)
- Make the `request` method only return a single message (this is the simpliest, but does limit you to more of a client-server type pattern)
- Maybe something in between? I can't quite think of one, but I don't have a monopoly on thoughts around messaging 😅

Reply:
> Make the request method only return a single message (this is the simplest, but does limit you to more of a client-server type pattern)

I think this would just cause people to make wrappers around the single reply mechanism for multiple-reply, so I'd avoid it.

Now, going over your other suggestion, I think the main question we should be asking is: What if the implementor doesn't really offer support for said feature? How should the implementor behave if, regardless, the field is set? And I'm not just saying this about `expected-replies`, but other `opts` too like `timeout-ms`. Rather than having options all together, I'm debating if we wouldn't be better off w/ entirely different functions, so that an implementor can make an explicit choice about implementing them or not. What are your thoughts, @thomastaylor312 ?

Reply:
If people wanted to opt out of implementations, each one of those functions would have to be a separate interface, which could get messy. I also feel like that no matter where each of those functions live, it would start to get really annoying to have to find the right function for each thing you're doing, but that is more of a minor gripe to me than the former. So my initial thought is that I don't quite think that would work.

The comment about options is interesting. A timeout value is simple enough that even if a client doesn't natively support it, it is really easy to add that via code. So I feel like that is fairly easy. I could say the same about expected replies, but I can also start to see the slippery slope here. At what point do additional options cross over into "yeah that is going to require custom logic for every implementation." Anyway, if my first answer is correct in that many functions would be a bad idea (which it could not be), then the best path forward would be to use options but to add even more documentation that we must be very careful adding anything else to it